scarlet/display.py

```diff
@@ -1,224 +1,268 @@
-    # We need to map this to a uint8 in the range [0,255]
-    rgb *= 255
-    rgb[rgb < 0] = 0
-    rgb[rgb > 255] = 255
-    rgb = rgb.astype(np.uint8)
-    # Now we move the SED to the outer axis and use only the selected colors
-    rgb = np.transpose(rgb[filter_indices], axes=(1, 2, 0))
+
     return rgb
 
+def channel_to_rgb(channels, norm):
+    """Convert RGB channel values to [0..255]
 
-def zscale(img, contrast=0.25, samples=500, fraction=.75):
-    """Calculate minimum and maximum pixel values based on the image
+    If normalized values are outside of the range [0..255], they will be
+    truncated such as to preserve the corresponding color.
+
+    Parameters
+    ----------
+    channels: array_like
+        Converted RGB channel values with dimension (3, height, width)
+    norm: `scarlet.display.Norm`
+        Norm to use for mapping in the allowed range [0..255]
+
+    Returns
+    -------
+    rgb: numpy array with dimensions (3, height, width) and dtype uint8
     """
-    if img.size > samples:
-        sorted_img = np.sort(np.random.choice(img.reshape(-1), size=samples))
-    else:
-        sorted_img = np.sort(img.reshape(-1))
+    rgb = norm(channels)
 
-    n_pix = len(sorted_img)
-    center = n_pix/2
-    median = sorted_img[int(center)]
-    idx = np.arange(n_pix)
+    # truncate at 0 and 1
+    rgb[rgb < 0] = 0 # looks better but is inaccurate because colors are not treated the same
+    # normalizing by a uniform maximum is critical for the color of saturated regions
+    rgb_max = rgb.max(axis=0)
+    rgb[:, rgb_max > 1] /= rgb_max[rgb_max > 1]
 
-    radius = 1-fraction/2
-    sample_slice = slice(int(np.floor((0.5-radius)*n_pix)), int(np.ceil((0.5+radius)*n_pix)))
-    result = np.polyfit(idx[sample_slice], sorted_img[sample_slice], 1)
-    slope, intercept = result
+    maxRGB = 255
+    return np.dstack((rgb * maxRGB)).astype(np.uint8)
 
-    z1 = max(median - center * (slope/contrast), np.min(img))
-    z2 = min(median + (n_pix-center-1) * (slope/contrast), np.max(img))
-    return z1, z2
+def rgb(img, filter_weights=None, fill_value=0, norm=None):
+    """Convert images to normalized RGB.
+
+    If normalized values are outside of the range [0..255], they will be
+    truncated such as to preserve the corresponding color.
+
+    Parameters
+    ----------
+    img: array_like
+        This should be an array with dimensions (bands, height, width).
+    filter_weights: array_like
+        Linear mapping with dimensions (channels, bands)
+    fill_value: float, default=`0`
+        Value to use for any masked pixels.
+    norm: `scarlet.display.Norm`, default `None`
+        Norm to use for mapping in the allowed range [0..255]. If `norm=None`,
+        `scarlet.display.LinearPercentileNorm` will be used.
+
+    Returns
+    -------
+    rgb: numpy array with dimensions (3, height, width) and dtype uint8
+    """
+    RGB = img_to_channel(img)
+    if norm is None:
```

Comment:
Why not just make the default value for `norm=LinearPercentileNorm`? The only time you should need to use `None` in the function signature is if the default value is an iterable or has a more complicated default value (like `filter_weights`).

Reply:
That's because you don't need the class of the norm, but an actual instance. The use case is if you define a survey-wide norm, linear or asinh, which you then need to pass to this method. As is, the method instantiates the norm if it is `None`, but it could not or should not do that all the time.

Reply:
You're right, I didn't catch that.